crypto/fipsmodule/rsa/rsa_impl.c

```diff
@@ -113,6 +108,27 @@ int rsa_check_public_key(const RSA *rsa) {
     OPENSSL_PUT_ERROR(RSA, RSA_R_KEY_SIZE_TOO_SMALL);
     return 0;
   }
+
+  return 1;
+}
+
+rsa_key_type_t detect_rsa_key_type(const RSA *rsa) {
+  if (rsa->d == NULL || BN_is_zero(rsa->d)) {
+      return RSA_KEY_PUBLIC;
+  }
+  if (rsa->e == NULL || BN_is_zero(rsa->e)) {
+      return RSA_KEY_PRIVATE_STRIPPED;
+  }
+  return RSA_KEY_PRIVATE_CRT;
+}
+
+int rsa_check_public_key(const RSA *rsa) {
+  if (rsa->n == NULL || rsa->e == NULL) {
+    OPENSSL_PUT_ERROR(RSA, RSA_R_VALUE_MISSING);
+    return 0;
+  }
+
+  rsa_check_key_common(rsa, RSA_KEY_PUBLIC);
```

Comment:
ah, shoot. yes i missed the ret val check here. tests still passed because the second call to `rsa_check_key_common` in `rsa_check_key_internal` caught issues with such keys, because as you point out we're calling `common` twice for public keys.

i'm going to rework this such that we call `common` exactly once for each key type.